### PR TITLE
Autoconfiguration for Steeltoe components via assembly detection and bootstrapping

### PR DESCRIPTION
**Pull request by macsux**
Applies the same principles as SpringBoot auto-config by automatically registering Steeltoe components based on the presence of necessary assemblies in the assembly loader search paths. This simplifies configuration of all Steeltoe components to just a a single `.AddSteeltoe()` on `IHostBuilder`.

See demo project https://github.com/macsux/steeltoe-autoconfig-demo

### CODE REVIEW

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                hostBuilder.WireIfAnyLoaded(WirePostgresConnection, PostgreSqlTypeLocator.Assemblies);
+                hostBuilder.WireIfAnyLoaded(WireRabbitMqConnection, RabbitMQTypeLocator.Assemblies);
+                hostBuilder.WireIfAnyLoaded(WireRedisConnectionMultiplexer, RedisTypeLocator.StackExchangeAssemblies);
+                hostBuilder.WireIfAnyLoaded(WireDistributedRedisCache, RedisTypeLocator.MicrosoftAssemblies.Except(new[] { "Microsoft.Extensions.Caching.Abstractions" }).ToArray());
+                hostBuilder.WireIfAnyLoaded(WireSqlServerConnection, SqlServerTypeLocator.Assemblies);
+            }
+
+            hostBuilder.WireIfLoaded(WireKubernetesActuators, SteeltoeAssemblies.Steeltoe_Management_KubernetesCore);
+            hostBuilder.WireIfLoaded(WireDiscoveryClient, SteeltoeAssemblies.Steeltoe_Discovery_ClientCore);
+            hostBuilder.WireIfLoaded(WireCloudFoundryActuators, SteeltoeAssemblies.Steeltoe_Management_CloudFoundryCore);
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:112]: `Management.CloudFoundryCore` isn't needed anymore, it's all obsolete and all functionality is contained in EndpointCore

File: src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    <ProjectReference Include="..\..\..\Security\src\Authentication.CloudFoundryCore\Steeltoe.Security.Authentication.CloudFoundryCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Security\src\Authentication.MtlsCore\Steeltoe.Security.Authentication.MtlsCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Security\src\DataProtection.CredHubCore\Steeltoe.Security.DataProtection.CredHubCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Security\src\DataProtection.RedisCore\Steeltoe.Security.DataProtection.RedisCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Connectors\src\CloudFoundry\Steeltoe.Connector.CloudFoundry.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Connectors\src\Connector.EFCore\Steeltoe.Connector.EFCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Logging\src\DynamicLogger\Steeltoe.Extensions.Logging.DynamicLogger.csproj" PrivateAssets="All" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\..\..\Connectors\src\ConnectorBase\**\*TypeLocator.cs" LinkBase=".\TypeLocators" />
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj:39]:
```suggestion
    <Compile Include="..\..\..\Connectors\src\ConnectorBase\**\*TypeLocator.cs" LinkBase="TypeLocators" />
```

Not sure if this matters for non-Windows+VS scenarios, but the `.\` adds an extra folder named `.` for me which adds an unnecessary layer

File: src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj [hunk 3]
```diff
@@ -0,0 +1,60 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <TargetFrameworks>netcoreapp3.1;net5.0</TargetFrameworks>
+    <RootNamespace>Steeltoe.Bootstrap.Autoconfig</RootNamespace>
+    <Description>Steeltoe autoconfiguration</Description>
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj:6]:
```suggestion
    <Description>**EXPERIMENTAL** Package for automatically configuring Steeltoe packages that have separately been added to a project.</Description>
```

File: src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    <ProjectReference Include="..\..\..\Configuration\src\ConfigServerCore\Steeltoe.Extensions.Configuration.ConfigServerCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Configuration\src\KubernetesCore\Steeltoe.Extensions.Configuration.KubernetesCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Configuration\src\PlaceholderCore\Steeltoe.Extensions.Configuration.PlaceholderCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Connectors\src\Connector.EF6Core\Steeltoe.Connector.EF6Core.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Connectors\src\ConnectorCore\Steeltoe.Connector.ConnectorCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Discovery\src\ClientCore\Steeltoe.Discovery.ClientCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Logging\src\DynamicSerilogCore\Steeltoe.Extensions.Logging.DynamicSerilogCore.csproj" PrivateAssets="All" />
+    <ProjectReference Include="..\..\..\Management\src\CloudFoundryCore\Steeltoe.Management.CloudFoundryCore.csproj" PrivateAssets="All" />
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/Steeltoe.Bootstrap.Autoconfig.csproj:25]:
```suggestion
```

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            {
+                hostBuilder.WireIfLoaded(WireCloudFoundryConfiguration, SteeltoeAssemblies.Steeltoe_Extensions_Configuration_CloudFoundryCore);
+            }
+
+            if (IsAssemblyLoaded(SteeltoeAssemblies.Steeltoe_Extensions_Configuration_KubernetesCore) && IsRunningInKubernetes())
+            {
+                WireKubernetesConfiguration(hostBuilder);
+            }
+
+            hostBuilder.WireIfLoaded(WirePlaceholderResolver, SteeltoeAssemblies.Steeltoe_Extensions_Configuration_PlaceholderCore);
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:96]: I don't see RandomValues, maybe we should include them before Placeholders?

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:97]: @hananiel should we autowire the SpringBoot config provider too? Which place in the order?

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+            }
+
+            return false;
+        }
+
+        // ReSharper disable once S1144
+        // ReSharper disable once UnusedMember.Local
+        private static bool IsRunningInCloudFoundry() => Environment.GetEnvironmentVariable("VCAP_APPLICATION") != null;
+
+        private static bool IsRunningInKubernetes() => Environment.GetEnvironmentVariable("KUBERNETES_SERVICE_HOST") != null;
```
Comment (TimHess) [inline src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:163]: `Platform.IsCloudFoundry` and `Platform.IsKubernetes` are already defined in [`Steeltoe.Common.Platform`](https://github.com/SteeltoeOSS/Steeltoe/blob/main/src/Common/src/Common/Platform.cs)

**Comment by TimHess** [comment]
Considerations not currently addressed in this work:

- Opt-out of autoconfiguration, ie: should we add some sort of exclusions list to `.AddSteeltoe` so that individual packages can be more specifically configured than what autoconfigure provides (eg: customizing the config provider order)
- Packages needing to be added to `IAppBuilder`, some of which need to be added in a specific order (eg: between AddRouting and `UseEndpoints`

Given we'll be distributing this as an experimental package, I think leaving out `IAppBuilder` extensions is OK given the combination of how complex it could get and how some of the current usage of these extensions could/should be moved into `IHostedService` classes instead. I think the idea of adding a way to exclude packages from autoconfiguration makes sense, but I'm open to discussion on whether it's needed for the first iteration of this package

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs (hunk 6 above)
Comment (TimHess) [reply src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:97]: Streams won't be included at this time, we can address the SpringBoot config provider later

File: src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs (hunk 1 above)
Comment (TimHess) [reply src/Bootstrap/src/Autoconfig/HostBuilderExtensions.cs:112]: Probably fine to leave this in for anybody that is still using it... we can drop it if/when the package is completely removed

**Review by hananiel — APPROVED** [review]

**Review by hananiel — APPROVED** [review]

**Pull request merged**